### PR TITLE
gateway-api: upgrade to v0.5.0-rc1

### PR DESCRIPTION
**Pull request by howardjohn**
This won't pass conformance tests currently due to some bugs in the test.

### CODE REVIEW

File: istioctl/pkg/util/handlers/handlers.go [hunk 1]
```diff
@@ -127,6 +128,12 @@ func SelectorsForObject(object runtime.Object) (namespace string, selector label
 		}
 		namespace = t.Namespace
 		selector, err = labels.Parse(gateway.GatewayNameLabel + "=" + t.Name)
+	case *gatewayapibeta.Gateway:
+		if !gateway.IsManagedBeta(&t.Spec) {
+			return "", nil, fmt.Errorf("gateway is not a managed gateway")
```
Comment (nolotz) [inline istioctl/pkg/util/handlers/handlers.go:133]: Consider using `errors.New()`

File: pilot/pkg/config/kube/gateway/conditions.go [hunk 2]
```diff
@@ -39,59 +39,79 @@ func createRouteStatus(gateways []routeParentReference, obj config.Config, curre
...
+	}
+	// Enhance error message when we have multiple
+	for k, gw := range seen {
+		if gw.DeniedReason == nil {
+			continue
 		}
-		if exist, f := seen[gw.OriginalReference]; !f || (exist.DeniedReason != nil && gw.DeniedReason == nil) {
-			seen[gw.OriginalReference] = gw
+		if failedCount[k] > 1 {
+			gw.DeniedReason.Message = fmt.Sprintf("failed to bind to %d parents, errors: %v", failedCount[k], gw.DeniedReason.Message)
```
Comment (nolotz) [inline pilot/pkg/config/kube/gateway/conditions.go:75]: Almost all Message's here start with a Capital letter, even those for ConfigError's. 
Is this here desired?

File: pilot/pkg/config/kube/gateway/conditions.go [hunk 3]
```diff
@@ -39,59 +39,79 @@ func createRouteStatus(gateways []routeParentReference, obj config.Config, curre
...
+		}
+		exist, f := seen[incoming.OriginalReference]
+		if !f {
+			exist = incoming
+		} else {
+			if incoming.DeniedReason == nil {
+				// The incoming gateway has no error; wipe out any errors
+				// When we have multiple, this means we attached without sectionName - we only need one success to be valid.
+				exist.DeniedReason = nil
+			} else if exist.DeniedReason != nil {
```
Comment (nolotz) [inline pilot/pkg/config/kube/gateway/conditions.go:62]: With `istiomultierror.New()` you could also handle reason.

File: pilot/pkg/config/kube/gateway/conditions.go [hunk 4]
```diff
@@ -39,59 +39,79 @@ func createRouteStatus(gateways []routeParentReference, obj config.Config, curre
...
 		// if *any* section has no errors, we will declare Admitted
-		if gw.DeniedReason != nil {
-			failedCount[gw.OriginalReference]++
+		if incoming.DeniedReason != nil {
+			failedCount[incoming.OriginalReference]++
+		} else {
+			successCount[incoming.OriginalReference]++
+		}
+		exist, f := seen[incoming.OriginalReference]
+		if !f {
```
Comment (nolotz) [inline pilot/pkg/config/kube/gateway/conditions.go:55]: You could `switch` this whole block away.

File: pilot/pkg/config/kube/gateway/conditions.go [hunk 5]
```diff
@@ -117,6 +146,19 @@ const (
 	InvalidConfiguration ConfigErrorReason = "InvalidConfiguration"
 )
 
+func NewConfigError(reason ConfigErrorReason, format string, args ...interface{}) *ConfigError {
```
Comment (nolotz) [inline pilot/pkg/config/kube/gateway/conditions.go:149]: Are there plans for using it? Looks like dead code.

**Review by costinm — APPROVED** [review]

File: pilot/pkg/config/kube/gateway/conditions.go (hunk 3 above)
Comment (costinm) [reply pilot/pkg/config/kube/gateway/conditions.go:62]: Not sure how - I think this needs to go into the status, as message string, not an issue of returning an error in code.

**Review by hanxiaop — APPROVED** [review]

**Pull request merged**